archive/100-theorems-list/83_friendship_graphs.lean

```diff
@@ -0,0 +1,285 @@
+  have dpos : 0 < d := by linarith,
+  have d_cast : ↑(d - 1) = (d : ℤ) - 1 := by norm_cast,
+  haveI : fact p.prime := nat.min_fac_prime (by linarith),
+  have hp2 : 2 ≤ p, { apply nat.prime.two_le, assumption },
+  have dmod : (d : zmod p) = 1,
+  { rw [← nat.succ_pred_eq_of_pos dpos, nat.succ_eq_add_one, nat.pred_eq_sub_one],
+    simp only [add_left_eq_self, nat.cast_add, nat.cast_one], apply p_dvd_d_pred, },
+  have Vmod := friendship_reg_card_mod_p hG dmod hd,
+  -- now we reduce to a trace calculation
+  have := zmod.trace_pow_card (G.adj_matrix (zmod p)),
+  contrapose! this, clear this,
+  -- the trace is 0 mod p when computed one way
+  rw [trace_adj_matrix, zero_pow],
+  -- but the trace is 1 mod p when computed the other way
+  rw friendship_reg_adj_pow_mod_p hG dmod hd hp2,
+  swap, { apply nat.prime.pos, assumption, },
+  { dunfold fintype.card at Vmod,
+    simp only [matrix.trace, diag_apply, mul_one, nsmul_eq_mul, linear_map.coe_mk, sum_const],
+    rw [Vmod, ← nat.cast_one, zmod.nat_coe_zmod_eq_zero_iff_dvd, nat.dvd_one, nat.min_fac_eq_one_iff],
+    linarith, },
+end
+
+lemma deg_le_one_friendship_has_pol (hd : G.is_regular_of_degree d) (hd1 : d ≤ 1) :
+  exists_politician G :=
+begin
+  have sq : d * d = d := by { interval_cases d; norm_num },
+  have h := friendship_reg_card hG hd,
+  rw sq at h,
+  have : fintype.card V ≤ 1,
+  { cases fintype.card V,
+    { exact zero_le _, },
+    { have : n = 0,
+      { rw [nat.succ_sub_succ_eq_sub, nat.sub_zero] at h,
+        linarith },
+      subst n, } },
+  use arbitrary V, intros w h,
+  exfalso, apply h,
+  apply fintype.card_le_one_iff.mp this,
+end
+
+lemma deg_two_neighbor_finset_eq (hd : G.is_regular_of_degree 2) (v : V) :
+  G.neighbor_finset v = finset.univ.erase v :=
+begin
+  apply finset.eq_of_subset_of_card_le,
+  { rw finset.subset_iff,
+    intro x,
+    rw [mem_neighbor_finset, finset.mem_erase],
+    exact λ h, ⟨ne.symm (G.ne_of_adj h), finset.mem_univ _⟩ },
+  convert_to 2 ≤ _,
+  convert_to _ = fintype.card V - 1,
+  { have hfr:= friendship_reg_card hG hd,
+    linarith },
+  { exact finset.card_erase_of_mem (finset.mem_univ _), },
+  dsimp [is_regular_of_degree, degree] at hd,
+  rw hd,
+end
+
+lemma deg_two_friendship_has_pol (hd : G.is_regular_of_degree 2) :
+  exists_politician G :=
+begin
+  have v := arbitrary V,
+  use v, intros w hvw,
+  rw [← mem_neighbor_finset, deg_two_neighbor_finset_eq hG hd v, finset.mem_erase],
+  exact ⟨ne.symm hvw, finset.mem_univ _⟩,
+end
+
+lemma deg_le_two_friendship_has_pol (hd : G.is_regular_of_degree d) (h : d ≤ 2) :
+  exists_politician G :=
+begin
+  interval_cases d,
+  iterate 2 { apply deg_le_one_friendship_has_pol hG hd, norm_num },
+  { exact deg_two_friendship_has_pol hG hd },
+end
+
+theorem friendship_theorem : exists_politician G :=
+begin
+  by_contradiction npG,
+  rcases counter_reg hG npG with ⟨d, dreg⟩,
+  have : d ≤ 2 ∨ 3 ≤ d := by omega,
+  cases this,
```

Comment:
If you refer to the names of the hypotheses generated by `cases` (both `this` here), you should give them explicitly. Here, `cases this with this this`, which I don't write as a suggestion because it sounds silly!